src/main/java/com/intershop/version/semantic/VersionResolver.java

```diff
@@ -0,0 +1,245 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class VersionResolver implements SemanticVersionResolver
+{
+    private static final Pattern DECIMAL_NONDECIMAL_SPLIT = Pattern.compile("^(\\d+)(\\D+)");
+    private static final Pattern NONDECIMAL_DECIMAL_SPLIT = Pattern.compile("^(\\D+)(\\d+)");
+    private static final Map<String, ExtensionType> MAP_EXTENSION_TO_TYPE = new HashMap<>();
+    static
+    {
+        MAP_EXTENSION_TO_TYPE.put("local", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("snapshot", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("a", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("alpha", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("b", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("beta", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("dev", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("m", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("milestone", ExtensionType.DEV);
+        MAP_EXTENSION_TO_TYPE.put("rc", ExtensionType.PRE);
+        MAP_EXTENSION_TO_TYPE.put("cr", ExtensionType.PRE);
+        MAP_EXTENSION_TO_TYPE.put("ea", ExtensionType.PRE);
+        MAP_EXTENSION_TO_TYPE.put("preview", ExtensionType.PRE);
+        MAP_EXTENSION_TO_TYPE.put("release", ExtensionType.GA);
+        MAP_EXTENSION_TO_TYPE.put("+", ExtensionType.BUILD);
+        MAP_EXTENSION_TO_TYPE.put("v", ExtensionType.BUILD);
+        MAP_EXTENSION_TO_TYPE.put("final", ExtensionType.GA);
+        MAP_EXTENSION_TO_TYPE.put("ga", ExtensionType.GA);
+        MAP_EXTENSION_TO_TYPE.put("jre", ExtensionType.PLATFORM);
+        MAP_EXTENSION_TO_TYPE.put("sp", ExtensionType.POST);
+    }
+    private static final Map<String, String> MAP_EXTENSION_TO_ITEM_EXTENSION = new HashMap<>();
+    // artificial number to identify build numbers or dates as increment, which are not incrementable
+    private static final Integer MAX_SEMANTIC_NUMBER = 10_000;
+    static
+    {
+        // ga alias
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("release", "");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("final", "");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("ga", "");
+        // build or version extensions
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("v", "");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("+", "");
+        // dev release alias
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("a", "alpha");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("b", "beta");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("m", "milestone");
+        // pre release alias
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("cr", "rc");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("preview", "rc");
+        MAP_EXTENSION_TO_ITEM_EXTENSION.put("ea", "rc");
+    }
+
+    public SemanticVersion apply(String version)
```

Comment:
I would name this method 'resolve'. 'apply' always sound like changing an internal state.

Reply:
than you can't use it as Function. this method "requires" an override annotation.